File: Minimum Size Subarray Sum - Leetcode 209/Minimum Size Subarray Sum - Leetcode 209.cpp
#include <vector>
#include <algorithm>
#include <climits>

class Solution {
public:
    int minSubArrayLen(int target, std::vector<int>& nums) {
        int minLength = INT_MAX;
        int sum = 0;
        int l = 0;

        for (int r = 0; r < nums.size(); ++r) {
            sum += nums[r];
            while (sum >= target) {
                minLength = std::min(minLength, r - l + 1);
                sum -= nums[l++];
            }
        }

        return minLength == INT_MAX ? 0 : minLength;
    }
};

File: Minimum Size Subarray Sum - Leetcode 209/Minimum Size Subarray Sum - Leetcode 209.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        min_length = float('inf')
        summ = 0
        l = 0
      
        for r in range(len(nums)):
            summ += nums[r]
            while summ >= target:
                min_length = min(min_length, r-l+1)
                summ -= nums[l]
                l += 1
        
        return min_length if min_length < float('inf') else 0
        # Time: O(n)
        # Space: O(1)
